tests/rules/test_asterisk_required.py

```diff
@@ -0,0 +1,199 @@
+
+
+def test_function_missing_asterisk():
+    source_code = """def my_function(a):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 1
+    assert occurrences[0] == Occurrence(
+        filename="my_file.py",
+        line_number=1,
+        rule_id=AsteriskRequiredRule.RULE_ID,
+        rule_label=AsteriskRequiredRule.RULE_LABEL,
+        function_name="my_function",
+    )
+
+
+def test_function_asterisk_too_late():
+    source_code = """def my_function(a, *, b):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 1
+    assert occurrences[0] == Occurrence(
+        filename="my_file.py",
+        line_number=1,
+        rule_id=AsteriskRequiredRule.RULE_ID,
+        rule_label=AsteriskRequiredRule.RULE_LABEL,
+        function_name="my_function",
+    )
+
+
+def test_function_arg_with_defaults():
+    source_code = """def my_function(a=42):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 1
+    assert occurrences[0] == Occurrence(
+        filename="my_file.py",
+        line_number=1,
+        rule_id=AsteriskRequiredRule.RULE_ID,
+        rule_label=AsteriskRequiredRule.RULE_LABEL,
+        function_name="my_function",
+    )
+
+
+def test_async_function_has_asterisk():
+    source_code = """async def my_function(*, a):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 0
+
+
+def test_async_function_missing_asterisk():
+    source_code = """async def my_function(a):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 1
+    assert occurrences[0] == Occurrence(
+        filename="my_file.py",
+        line_number=1,
+        rule_id=AsteriskRequiredRule.RULE_ID,
+        rule_label=AsteriskRequiredRule.RULE_LABEL,
+        function_name="my_function",
+    )
+
+
+def test_self_outside_of_class_not_matched():
+    source_code = """def my_function(self):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 0
+
+
+def test_cls_outside_of_class_not_matched():
+    source_code = """def my_function(cls):
+        pass
+    """
+
+    rule = AsteriskRequiredRule(filename="my_file.py", source_code=source_code)
+    occurrences = rule.check(source_code=source_code)
+
+    assert len(occurrences) == 0
```

Comment:
suggestion: Die ganzen gleichen Tests, die nichts finden sollen, könntest Du mit parametrize verkürzen: https://docs.pytest.org/en/latest/example/parametrize.html

Reply:
Finde, man verliert dann den Kontext des Tests.

Reply:
Sehe ich anders:
```
@pytest.mark.parametrize(
    ("test_code", "rule"),
    [
        (
            # test_self_outside_of_class_not_matched
            """def my_function(self):
            pass
        """,
            AsteriskRequiredRule,
        ),
        (
            # test_cls_outside_of_class_not_matched
            """def my_function(cls):
            pass
        """,
            AsteriskRequiredRule,
        ),
    ]
)
def test_rules(test_code, rule):
    rule = rule(filename="my_file.py", source_code=test_code)

    occurrences = rule.check(source_code=test_code)

    assert len(occurrences) == 0
```

Wenn er was finden soll, dann könntest Du ja sogar den erwarteten Fehlertext mit in die Parameter setzen.

Aber letztendlich ist es mir egal. Geschmackssache.